File: test/pyssz/test_progressive_container.py
#!/usr/bin/env python3

import sys
import os
# Add local remerkleable to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), 'remerkleable'))

from remerkleable.core import View
from remerkleable.basic import uint16, uint64
from remerkleable.byte_arrays import Bytes32
from remerkleable.complex import Container
from remerkleable.progressive import ProgressiveContainer


# Define a base container with all possible fields
class TestBase(Container):
    a: uint64        # index 0
    b: uint64        # index 1  
    c: Bytes32       # index 2
    d: uint16        # index 3
    e: uint64        # index 4
    f: Bytes32       # index 5
    g: uint16        # index 6
    h: uint64        # index 7


def main():
    # Test case 1: Enable fields at indices 0, 1, 3, 5, 7 (skip 2, 4, 6)
    # Active fields: [1, 1, 0, 1, 0, 1, 0, 1]
    print("=== Progressive Container Test ===")
    
    # Create a progressive container class with specific active fields
    active_fields = [1, 1, 0, 1, 0, 1, 0, 1]
    TestProgressive = ProgressiveContainer(active_fields)
    
    # Define the container structure dynamically
    class TestStruct(TestProgressive):
        a: uint64        # active (index 0)
        b: uint64        # active (index 1)  
        d: uint16        # active (index 3)
        f: Bytes32       # active (index 5)
        h: uint64        # active (index 7)
    
    # Create test data
    test_data = TestStruct(
        a=uint64(12345),
        b=uint64(67890),
        d=uint16(999),
        f=Bytes32(b'\x11' * 32),
        h=uint64(0xdeadbeef)
    )
    
    print(f"Active fields: a={test_data.a}, b={test_data.b}, d={test_data.d}, f={test_data.f.hex()}, h={test_data.h}")
    print(f"Active field pattern: {active_fields}")
    
    # Get SSZ encoding
    try:
        ssz_encoded = test_data.encode_bytes()
        print(f"\nSSZ encoded: {ssz_encoded.hex()}")
        print(f"SSZ length: {len(ssz_encoded)} bytes")
    except Exception as e:
        print(f"\nSSZ encoding error: {e}")
        import traceback
        traceback.print_exc()
    
    # Get hash tree root
    try:
        tree_root = test_data.hash_tree_root()
        print(f"\nTree root: {tree_root.hex()}")
    except Exception as e:
        print(f"\nTree root error: {e}")
        import traceback
        traceback.print_exc()
    
    # Test case 2: Different active fields pattern
    print("\n=== Test Case 2: Different Active Fields ===")
    active_fields2 = [1, 0, 1, 0, 1, 0, 1]  # indices 0, 2, 4, 6
    TestProgressive2 = ProgressiveContainer(active_fields2)
    
    class TestStruct2(TestProgressive2):
        a: uint64        # active (index 0)
        c: Bytes32       # active (index 2)
        e: uint64        # active (index 4)
        g: uint16        # active (index 6)
    
    test_data2 = TestStruct2(
        a=uint64(11111),
        c=Bytes32(b'\x22' * 32),
        e=uint64(33333),
        g=uint16(444)
    )
    
    print(f"Active fields: a={test_data2.a}, c={test_data2.c.hex()}, e={test_data2.e}, g={test_data2.g}")
    print(f"Active field pattern: {active_fields2}")
    
    try:
        ssz_encoded2 = test_data2.encode_bytes()
        print(f"\nSSZ encoded: {ssz_encoded2.hex()}")
        print(f"SSZ length: {len(ssz_encoded2)} bytes")
        
        tree_root2 = test_data2.hash_tree_root()
        print(f"Tree root: {tree_root2.hex()}")
    except Exception as e:
        print(f"\nError in test case 2: {e}")
        import traceback
        traceback.print_exc()
    
    # Test case 3: Minimal case - only first few fields 
    print("\n=== Test Case 3: Only First Few Fields ===")
    active_fields3 = [1, 1, 1]  # Only first 3 fields active
    TestProgressive3 = ProgressiveContainer(active_fields3)
    
    class TestStruct3(TestProgressive3):
        a: uint64        # active (index 0)
        b: uint64        # active (index 1)
        c: Bytes32       # active (index 2)
    
    test_data3 = TestStruct3(
        a=uint64(99999),
        b=uint64(88888),
        c=Bytes32(b'\x33' * 32)
    )
    
    print(f"Active fields: a={test_data3.a}, b={test_data3.b}, c={test_data3.c.hex()}")
    print(f"Active field pattern: {active_fields3}")
    
    try:
        ssz_encoded3 = test_data3.encode_bytes()
        print(f"\nSSZ encoded: {ssz_encoded3.hex()}")
        print(f"SSZ length: {len(ssz_encoded3)} bytes")
        
        tree_root3 = test_data3.hash_tree_root()
        print(f"Tree root: {tree_root3.hex()}")
    except Exception as e:
        print(f"\nError in test case 3: {e}")
        import traceback
        traceback.print_exc()
    
    # Output results in a format we can use for Go comparison
    print("\n=== Results Summary for Go Implementation ===")
    print("Test Case 1:")
    print(f"  Active: [1, 1, 0, 1, 0, 1, 0, 1]")
    print(f"  SSZ: {ssz_encoded.hex() if 'ssz_encoded' in locals() else 'ERROR'}")
    print(f"  Root: {tree_root.hex() if 'tree_root' in locals() else 'ERROR'}")
    print("\nTest Case 2:")
    print(f"  Active: [1, 0, 1, 0, 1, 0, 1]")
    print(f"  SSZ: {ssz_encoded2.hex() if 'ssz_encoded2' in locals() else 'ERROR'}")
    print(f"  Root: {tree_root2.hex() if 'tree_root2' in locals() else 'ERROR'}")
    print("\nTest Case 3:")
    print(f"  Active: [1, 1, 1]")
    print(f"  SSZ: {ssz_encoded3.hex() if 'ssz_encoded3' in locals() else 'ERROR'}")
    print(f"  Root: {tree_root3.hex() if 'tree_root3' in locals() else 'ERROR'}")


if __name__ == "__main__":
    main()

File: test/pyssz/test_progressive_container.go
// Copyright (c) 2025 pk910
// SPDX-License-Identifier: Apache-2.0

package main

import (
	"encoding/hex"
	"fmt"

	dynssz "github.com/pk910/dynamic-ssz"
)

// go run test_progressive_container.go

// Test case 1: active fields at indices 0, 1, 3, 5, 7
type TestProgressive1 struct {
	A uint64   `ssz-index:"0"`
	B uint64   `ssz-index:"1"`
	D uint16   `ssz-index:"3"`
	F [32]byte `ssz-index:"5"`
	H uint64   `ssz-index:"7"`
}

// Test case 2: active fields at indices 0, 2, 4, 6
type TestProgressive2 struct {
	A uint64   `ssz-index:"0"`
	C [32]byte `ssz-index:"2"`
	E uint64   `ssz-index:"4"`
	G uint16   `ssz-index:"6"`
}

// Test case 3: active fields at indices 0, 1, 2
type TestProgressive3 struct {
	A uint64   `ssz-index:"0"`
	B uint64   `ssz-index:"1"`
	C [32]byte `ssz-index:"2"`
}

func main() {
	ds := dynssz.NewDynSsz(nil)

	fmt.Println("=== Progressive Container Test (Go) ===")

	// Test case 1: active fields at indices 0, 1, 3, 5, 7
	fmt.Println("\nTest Case 1:")
	fmt.Println("Active field indices: 0, 1, 3, 5, 7")

	var fBytes1 [32]byte
	for i := 0; i < 32; i++ {
		fBytes1[i] = 0x11
	}

	test1 := TestProgressive1{
		A: 12345,
		B: 67890,
		D: 999,
		F: fBytes1,
		H: 0xdeadbeef,
	}

	ssz1, err := ds.MarshalSSZ(test1)
	if err != nil {
		fmt.Printf("SSZ encoding error: %v\n", err)
	} else {
		fmt.Printf("SSZ encoded: %x\n", ssz1)
		fmt.Printf("SSZ length: %d bytes\n", len(ssz1))
	}

	root1, err := ds.HashTreeRoot(test1)
	if err != nil {
		fmt.Printf("Hash tree root error: %v\n", err)
	} else {
		fmt.Printf("Tree root: %x\n", root1)
	}

	// Compare with Python results
	expectedSSZ1 := "39300000000000003209010000000000e7031111111111111111111111111111111111111111111111111111111111111111efbeadde00000000"
	expectedRoot1 := "816edda1de1b98f441d929ac0d5a4dfb08c059df22a22fae9add106dadf0364c"

	if hex.EncodeToString(ssz1) == expectedSSZ1 {
		fmt.Println("✓ SSZ matches Python implementation")
	} else {
		fmt.Println("✗ SSZ does not match Python implementation")
		fmt.Printf("  Expected: %s\n", expectedSSZ1)
		fmt.Printf("  Got:      %x\n", ssz1)
	}

	if hex.EncodeToString(root1[:]) == expectedRoot1 {
		fmt.Println("✓ Root matches Python implementation")
	} else {
		fmt.Println("✗ Root does not match Python implementation")
		fmt.Printf("  Expected: %s\n", expectedRoot1)
		fmt.Printf("  Got:      %x\n", root1)
	}

	// Test case 2: active fields at indices 0, 2, 4, 6
	fmt.Println("\nTest Case 2:")
	fmt.Println("Active field indices: 0, 2, 4, 6")

	var cBytes2 [32]byte
	for i := 0; i < 32; i++ {
		cBytes2[i] = 0x22
	}

	test2 := TestProgressive2{
		A: 11111,
		C: cBytes2,
		E: 33333,
		G: 444,
	}

	ssz2, err := ds.MarshalSSZ(test2)
	if err != nil {
		fmt.Printf("SSZ encoding error: %v\n", err)
	} else {
		fmt.Printf("SSZ encoded: %x\n", ssz2)
		fmt.Printf("SSZ length: %d bytes\n", len(ssz2))
	}

	root2, err := ds.HashTreeRoot(test2)
	if err != nil {
		fmt.Printf("Hash tree root error: %v\n", err)
	} else {
		fmt.Printf("Tree root: %x\n", root2)
	}

	// Compare with Python results
	expectedSSZ2 := "672b00000000000022222222222222222222222222222222222222222222222222222222222222223582000000000000bc01"
	expectedRoot2 := "51e5fe5ce5c39cee27fd091d3de9b73d90e43a3d8fcf85dd96baa22db0aa6aef"

	if hex.EncodeToString(ssz2) == expectedSSZ2 {
		fmt.Println("✓ SSZ matches Python implementation")
	} else {
		fmt.Println("✗ SSZ does not match Python implementation")
		fmt.Printf("  Expected: %s\n", expectedSSZ2)
		fmt.Printf("  Got:      %x\n", ssz2)
	}

	if hex.EncodeToString(root2[:]) == expectedRoot2 {
		fmt.Println("✓ Root matches Python implementation")
	} else {
		fmt.Println("✗ Root does not match Python implementation")
		fmt.Printf("  Expected: %s\n", expectedRoot2)
		fmt.Printf("  Got:      %x\n", root2)
	}

	// Test case 3: active fields at indices 0, 1, 2
	fmt.Println("\nTest Case 3:")
	fmt.Println("Active field indices: 0, 1, 2")

	var cBytes3 [32]byte
	for i := 0; i < 32; i++ {
		cBytes3[i] = 0x33
	}

	test3 := TestProgressive3{
		A: 99999,
		B: 88888,
		C: cBytes3,
	}

	ssz3, err := ds.MarshalSSZ(test3)
	if err != nil {
		fmt.Printf("SSZ encoding error: %v\n", err)
	} else {
		fmt.Printf("SSZ encoded: %x\n", ssz3)
		fmt.Printf("SSZ length: %d bytes\n", len(ssz3))
	}

	root3, err := ds.HashTreeRoot(test3)
	if err != nil {
		fmt.Printf("Hash tree root error: %v\n", err)
	} else {
		fmt.Printf("Tree root: %x\n", root3)
	}

	// Compare with Python results
	expectedSSZ3 := "9f86010000000000385b0100000000003333333333333333333333333333333333333333333333333333333333333333"
	expectedRoot3 := "882e46351726c479519a1a51d821f4acbe3cb99531a64eb6ff8c393b7676180c"

	if hex.EncodeToString(ssz3) == expectedSSZ3 {
		fmt.Println("✓ SSZ matches Python implementation")
	} else {
		fmt.Println("✗ SSZ does not match Python implementation")
		fmt.Printf("  Expected: %s\n", expectedSSZ3)
		fmt.Printf("  Got:      %x\n", ssz3)
	}

	if hex.EncodeToString(root3[:]) == expectedRoot3 {
		fmt.Println("✓ Root matches Python implementation")
	} else {
		fmt.Println("✗ Root does not match Python implementation")
		fmt.Printf("  Expected: %s\n", expectedRoot3)
		fmt.Printf("  Got:      %x\n", root3)
	}

	fmt.Println("\n=== Summary ===")
	fmt.Println("All test cases compare SSZ encoding and hash tree root")
	fmt.Println("between Go dynamic-ssz and Python remerkleable implementations")
	fmt.Println("for ProgressiveContainer with different active field patterns.")
}
